### PR TITLE
test: Add integration tests for GCSOutputStream and GcsFileSystemImplIntegrationTest for write implementation

### PR DESCRIPTION
**Pull request by dheerajsngh**
## Type of Change

- [ ] `feat`: A new feature
- [ ] `fix`: A bug fix
- [ ] `docs`: Documentation only changes
- [ ] `style`: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] `refactor`: A code change that neither fixes a bug nor adds a feature
- [ ] `perf`: A code change that improves performance
- [x] `test`: Adding missing tests or correcting existing tests
- [ ] `chore`: Changes to the build process or auxiliary tools and libraries such as documentation generation

### What?
- Added `GoogleCloudStorageOutputStreamIntegrationTest` to validate write operations, exception handling, and edge cases for the new core stream writer.
- Introduced `TestOutputStreamOutputFile` as a testing helper.
- Added new integration tests in `GcsFileSystemImplIntegrationTest` to verify file creation flows via `GcsWriteChannel`.
- Updated `IntegrationTestHelper` to support the new write testing scenarios.

### Why?
To ensure the correctness and stability of the newly integrated `GoogleCloudStorageOutputStream` and `GcsWriteChannel`. These tests are crucial for verifying that data is reliably written to GCS and that stream resources and exceptions are handled properly to prevent future regressions.

## Checklist

- [x] PR title follows [Conventional Commits](https://www.conventionalcommits.org/) (e.g., `feat(core): ...`)
- [x] All files include the Apache License 2.0 header
- [ ] Documentation has been updated to reflect changes

---
*Generated/Assisted by Agent? Yes*

### CODE REVIEW

File: core/src/integrationTest/java/com/google/cloud/gcs/analyticscore/core/TestOutputStreamOutputFile.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+      @Override
+      public void write(int b) throws IOException {
+        ByteBuffer buffer = ByteBuffer.allocate(1);
+        buffer.put((byte) b).flip();
+        while (buffer.hasRemaining()) {
+          int written = channel.write(buffer);
+          if (written < 0) throw new EOFException("Channel closed unexpectedly");
+          position += written;
+        }
```
[the remark this answers is not in the document]
Comment (dheerajsngh) [reply core/src/integrationTest/java/com/google/cloud/gcs/analyticscore/core/TestOutputStreamOutputFile.java]: Done.

File: core/src/integrationTest/java/com/google/cloud/gcs/analyticscore/core/GoogleCloudStorageOutputStreamIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,588 @@
...
+        try (GoogleCloudStorageOutputStream outputStream = GoogleCloudStorageOutputStream.create(customFs, itemId)) {
+          outputStream.write(TEST_CONTENT);
+        }
+        assertThat(IntegrationTestHelper.objectPresentInBucket(fileName)).isTrue();
+        assertThat(customFs.getFileInfo(uri).getItemInfo().getSize()).isEqualTo((long) TEST_CONTENT.length);
+      }
+    } finally {
+      Files.deleteIfExists(tempDir);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (dheerajsngh) [reply core/src/integrationTest/java/com/google/cloud/gcs/analyticscore/core/GoogleCloudStorageOutputStreamIntegrationTest.java]: Fixed.

File: core/src/integrationTest/java/com/google/cloud/gcs/analyticscore/core/GoogleCloudStorageOutputStreamIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,588 @@
...
+    URI uri = IntegrationTestHelper.getGcsObjectUriForFile(fileName);
+    BlobId blobId = BlobId.fromGsUtilUri(uri.toString());
+    blobsToDelete.add(blobId);
+    GcsItemId itemId = GcsItemId.builder().setBucketName(blobId.getBucket()).setObjectName(blobId.getName()).build();
+
+    GcsClientOptions clientOptions = GcsClientOptions.builder()
+        .setUploadType(GcsClientOptions.UploadType.PARALLEL_COMPOSITE_UPLOAD)
+        .setPcuBufferCount(2)
+        .setPcuBufferCapacity(16 * 1024 * 1024) // 16MB
+        .setPcuPartFileCleanupType(GcsClientOptions.PartFileCleanupType.ALWAYS)
```
[the remark this answers is not in the document]
Comment (dheerajsngh) [reply core/src/integrationTest/java/com/google/cloud/gcs/analyticscore/core/GoogleCloudStorageOutputStreamIntegrationTest.java]: Fixed.

File: core/src/integrationTest/java/com/google/cloud/gcs/analyticscore/core/GoogleCloudStorageOutputStreamIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,588 @@
...
+    blobsToDelete.add(blobId);
+    GcsItemId itemId = GcsItemId.builder().setBucketName(blobId.getBucket()).setObjectName(blobId.getName()).build();
+
+    GcsClientOptions clientOptions = GcsClientOptions.builder()
+        .setUploadChunkSize(256 * 1024)
+        .build();
+    GcsWriteOptions writeOptions = GcsWriteOptions.builder().build();
+
+    int totalSize = 1024 * 1024; // 1 MB total size
+    byte[] chunk = new byte[1024]; // 1 KB chunks written locally
```
[the remark this answers is not in the document]
Comment (dheerajsngh) [reply core/src/integrationTest/java/com/google/cloud/gcs/analyticscore/core/GoogleCloudStorageOutputStreamIntegrationTest.java]: Fixed.

File: core/src/integrationTest/java/com/google/cloud/gcs/analyticscore/core/GoogleCloudStorageOutputStreamIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,602 @@
...
+    try (ParquetReader<Group> reader = new GroupParquetReaderBuilder(inputFile).withConf(conf).build();
+        ParquetWriter<Group> writer = ExampleParquetWriter.builder(outputFile).withConf(conf).build()) {
+      Group group;
+      while ((group = reader.read()) != null) {
+        writer.write(group);
+        sourceRecordSignatures.add(group.toString());
+        recordsCopied++;
+        if (recordsCopied >= 100) break;
+      }
+    }
```
[the remark this answers is not in the document]
Comment (dheerajsngh) [reply core/src/integrationTest/java/com/google/cloud/gcs/analyticscore/core/GoogleCloudStorageOutputStreamIntegrationTest.java]: Fixed

File: core/src/integrationTest/java/com/google/cloud/gcs/analyticscore/core/GoogleCloudStorageOutputStreamIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,602 @@
...
+    Configuration conf = new Configuration();
+    GroupWriteSupport.setSchema(schema, conf);
+
+    try (ParquetWriter<Group> writer = ExampleParquetWriter.builder(outputFile)
+        .withConf(conf)
+        .build()) {
+      SimpleGroupFactory groupFactory = new SimpleGroupFactory(schema);
+      writer.write(groupFactory.newGroup().append(PARQUET_FIELD_NAME, PARQUET_VAL_ALICE));
+      writer.write(groupFactory.newGroup().append(PARQUET_FIELD_NAME, PARQUET_VAL_BOB));
+    }
```
[the remark this answers is not in the document]
Comment (dheerajsngh) [reply core/src/integrationTest/java/com/google/cloud/gcs/analyticscore/core/GoogleCloudStorageOutputStreamIntegrationTest.java]: Fixed

File: core/src/integrationTest/java/com/google/cloud/gcs/analyticscore/core/GoogleCloudStorageOutputStreamIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,602 @@
...
+    blobsToDelete.add(blobId);
+    GcsItemId itemId = GcsItemId.builder().setBucketName(blobId.getBucket()).setObjectName(blobId.getName()).build();
+
+    GcsClientOptions clientOptions = GcsClientOptions.builder()
+        .setUploadChunkSize(256 * KB)
+        .build();
+    GcsWriteOptions writeOptions = GcsWriteOptions.builder().build();
+
+    int totalSize = 1 * MB;
+    byte[] chunk = new byte[1 * KB];
```
[the remark this answers is not in the document]
Comment (dheerajsngh) [reply core/src/integrationTest/java/com/google/cloud/gcs/analyticscore/core/GoogleCloudStorageOutputStreamIntegrationTest.java]: Fixed

File: client/src/integrationTest/java/com/google/cloud/gcs/analyticscore/client/GcsFileSystemImplIntegrationTest.java [hunk 8]
```diff
@@ -115,4 +120,119 @@ public void getFileInfo_anonymousCredentialProvided_urlPointsToPrivateObject_thr
...
+                .setBucketName(bucketName)
+                .setObjectName(objectName)
+                .build();
+
+        GcsWriteOptions writeOptions = GcsWriteOptions.builder().build();
+
+        assertThrows(FileNotFoundException.class, () -> {
+            try (WritableByteChannel channel = gcsFileSystem.create(itemId, writeOptions)) {
+                channel.write(ByteBuffer.wrap("test".getBytes(StandardCharsets.UTF_8)));
+            }
```
[the remark this answers is not in the document]
Comment (dheerajsngh) [reply client/src/integrationTest/java/com/google/cloud/gcs/analyticscore/client/GcsFileSystemImplIntegrationTest.java]: Fixed

File: client/src/integrationTest/java/com/google/cloud/gcs/analyticscore/client/GcsFileSystemImplIntegrationTest.java [hunk 9]
```diff
@@ -115,4 +120,119 @@ public void getFileInfo_anonymousCredentialProvided_urlPointsToPrivateObject_thr
...
+                .setBucketName(bucketName)
+                .setObjectName(objectName)
+                .build();
+
+        GcsWriteOptions writeOptions = GcsWriteOptions.builder().build();
+        byte[] content = "test content".getBytes(StandardCharsets.UTF_8);
+
+        try (WritableByteChannel channel = gcsFileSystem.create(itemId, writeOptions)) {
+            channel.write(ByteBuffer.wrap(content));
+        }
```
[the remark this answers is not in the document]
Comment (dheerajsngh) [reply client/src/integrationTest/java/com/google/cloud/gcs/analyticscore/client/GcsFileSystemImplIntegrationTest.java]: Fixed

File: client/src/integrationTest/java/com/google/cloud/gcs/analyticscore/client/GcsFileSystemImplIntegrationTest.java [hunk 10]
```diff
@@ -115,4 +120,119 @@ public void getFileInfo_anonymousCredentialProvided_urlPointsToPrivateObject_thr
...
+
+        GcsWriteOptions noOverwriteOptions = GcsWriteOptions.builder()
+                .setOverwriteExisting(false)
+                .build();
+
+        assertThrows(FileAlreadyExistsException.class, () -> {
+            try (WritableByteChannel channel = gcsFileSystem.create(itemId, noOverwriteOptions)) {
+                channel.write(ByteBuffer.wrap("test".getBytes(StandardCharsets.UTF_8)));
+            }
```
[the remark this answers is not in the document]
Comment (dheerajsngh) [reply client/src/integrationTest/java/com/google/cloud/gcs/analyticscore/client/GcsFileSystemImplIntegrationTest.java]: Fixed.

File: client/src/integrationTest/java/com/google/cloud/gcs/analyticscore/client/GcsFileSystemImplIntegrationTest.java [hunk 11]
```diff
@@ -115,4 +120,119 @@ public void getFileInfo_anonymousCredentialProvided_urlPointsToPrivateObject_thr
...
+        GcsItemId itemId = GcsItemId.builder()
+                .setBucketName(bucketName)
+                .setObjectName(objectName)
+                .build();
+
+        // We do a preliminary setup write
+        GcsWriteOptions writeOptions = GcsWriteOptions.builder().build();
+        try (WritableByteChannel channel = gcsFileSystem.create(itemId, writeOptions)) {
+            channel.write(ByteBuffer.wrap("test".getBytes(StandardCharsets.UTF_8)));
+        }
```
[the remark this answers is not in the document]
Comment (dheerajsngh) [reply client/src/integrationTest/java/com/google/cloud/gcs/analyticscore/client/GcsFileSystemImplIntegrationTest.java]: Fixed.

File: client/src/integrationTest/java/com/google/cloud/gcs/analyticscore/client/GcsFileSystemImplIntegrationTest.java [hunk 12]
```diff
@@ -115,4 +120,119 @@ public void getFileInfo_anonymousCredentialProvided_urlPointsToPrivateObject_thr
...
+                .setBucketName(bucketName)
+                .setObjectName(objectName)
+                .build();
+
+        GcsWriteOptions writeOptions = GcsWriteOptions.builder().build();
+        byte[] content = "test content".getBytes(StandardCharsets.UTF_8);
+
+        try (WritableByteChannel channel = gcsFileSystem.create(itemId, writeOptions)) {
+            channel.write(ByteBuffer.wrap(content));
+        }
```
[the remark this answers is not in the document]
Comment (dheerajsngh) [reply client/src/integrationTest/java/com/google/cloud/gcs/analyticscore/client/GcsFileSystemImplIntegrationTest.java]: Fixed.

File: core/src/integrationTest/java/com/google/cloud/gcs/analyticscore/core/GoogleCloudStorageOutputStreamIntegrationTest.java [hunk 13]
```diff
@@ -0,0 +1,602 @@
...
+              .forEach(path -> {
+                try {
+                  Files.deleteIfExists(path);
+                } catch (IOException e) {
+                  // Ignore cleanup exceptions to avoid masking the main test failure
+                }
+              });
+        }
+      }
+    }
```
[the remark this answers is not in the document]
Comment (dheerajsngh) [reply core/src/integrationTest/java/com/google/cloud/gcs/analyticscore/core/GoogleCloudStorageOutputStreamIntegrationTest.java]: Fixed.

**Review by prudhvimaharishi — DISMISSED** [review]

File: client/src/integrationTest/java/com/google/cloud/gcs/analyticscore/client/GcsFileSystemImplIntegrationTest.java [hunk 14]
```diff
@@ -115,4 +120,111 @@ public void getFileInfo_anonymousCredentialProvided_urlPointsToPrivateObject_thr
 
         assertThat(exception).hasMessageThat().contains("Unable to access blob");
     }
+
+    @Test
+    @EnabledIfSystemProperty(named = "gcs.integration.test.bucket", matches = ".+")
+    public void create_object_canWriteContent() throws IOException {
```
Comment (animesh-g) [inline client/src/integrationTest/java/com/google/cloud/gcs/analyticscore/client/GcsFileSystemImplIntegrationTest.java]: Tests in this file create objects but never clean them. We should delete them similar to other file via `blobsToDelete`.

File: core/src/integrationTest/java/com/google/cloud/gcs/analyticscore/core/GoogleCloudStorageOutputStreamIntegrationTest.java [hunk 15]
```diff
@@ -0,0 +1,523 @@
...
+    GcsClientOptions clientOptions = GcsClientOptions.builder()
+        .setUploadType(GcsClientOptions.UploadType.JOURNALING)
+        .setTemporaryPaths(Collections.singletonList("/tmp/dummy-path"))
+        .build();
+    assertThrows(UnsupportedOperationException.class, () -> createFileSystemWithClientOptions(clientOptions));
+  }
+
+  @Test
+  void writeToNonExistentBucket_throwsAccessDeniedException() {
+    BlobId blobId = BlobId.of(NON_EXISTENT_BUCKET, NON_EXISTENT_FILE);
```
Comment (animesh-g) [inline core/src/integrationTest/java/com/google/cloud/gcs/analyticscore/core/GoogleCloudStorageOutputStreamIntegrationTest.java]: We should append a uuid in NON_EXISTENT_BUCKET to keep it unique. Given it is a public repo, if someone creates a bucket with same name, this test will start failing.

File: core/src/integrationTest/java/com/google/cloud/gcs/analyticscore/core/GoogleCloudStorageOutputStreamIntegrationTest.java [hunk 16]
```diff
@@ -0,0 +1,523 @@
...
+  }
+
+  @AfterEach
+  void tearDown() {
+    if (IntegrationTestHelper.storage != null) {
+      for (BlobId blobId : blobsToDelete) {
+        try {
+          IntegrationTestHelper.storage.delete(blobId);
+        } catch (Exception e) {
+          // Ignore cleanup errors
```
Comment (animesh-g) [inline core/src/integrationTest/java/com/google/cloud/gcs/analyticscore/core/GoogleCloudStorageOutputStreamIntegrationTest.java]: Let's add a log here so that we can catch issues while debugging.

File: core/src/integrationTest/java/com/google/cloud/gcs/analyticscore/core/GoogleCloudStorageOutputStreamIntegrationTest.java [hunk 17]
```diff
@@ -0,0 +1,523 @@
...
+    try (GoogleCloudStorageOutputStream outputStream = GoogleCloudStorageOutputStream.create(createFileSystemWithWriteOptions(writeOptions), ctx.itemId)) {
+      outputStream.write(content);
+    }
+
+    assertThat(IntegrationTestHelper.objectPresentInBucket(ctx.fileName)).isTrue();
+    assertThat(gcsFileSystem.getFileInfo(ctx.uri).getItemInfo().getSize()).isEqualTo((long) content.length);
+  }
+
+  @Test
+  void write_overwriteDisabledAndFileExists_throwsFileAlreadyExistsException() throws IOException {
```
Comment (animesh-g) [inline core/src/integrationTest/java/com/google/cloud/gcs/analyticscore/core/GoogleCloudStorageOutputStreamIntegrationTest.java]: Isn't create_overwriteDisabled_throwsFileAlreadyExistsException already cover this? Applies to the other tests as well. I think there are multiple tests testing the same behavior at client and core packages.

Comment (dheerajsngh) [reply core/src/integrationTest/java/com/google/cloud/gcs/analyticscore/core/GoogleCloudStorageOutputStreamIntegrationTest.java]: Removed the duplicate tests from GCSOS integration test which are alreay present in GCSFileSystem integration tests.

File: core/src/integrationTest/java/com/google/cloud/gcs/analyticscore/core/GoogleCloudStorageOutputStreamIntegrationTest.java (hunk 16 above)
Comment (dheerajsngh) [reply core/src/integrationTest/java/com/google/cloud/gcs/analyticscore/core/GoogleCloudStorageOutputStreamIntegrationTest.java]: Done.

File: core/src/integrationTest/java/com/google/cloud/gcs/analyticscore/core/GoogleCloudStorageOutputStreamIntegrationTest.java (hunk 15 above)
Comment (dheerajsngh) [reply core/src/integrationTest/java/com/google/cloud/gcs/analyticscore/core/GoogleCloudStorageOutputStreamIntegrationTest.java]: Done.

File: client/src/integrationTest/java/com/google/cloud/gcs/analyticscore/client/GcsFileSystemImplIntegrationTest.java (hunk 14 above)
Comment (dheerajsngh) [reply client/src/integrationTest/java/com/google/cloud/gcs/analyticscore/client/GcsFileSystemImplIntegrationTest.java]: Done.

**Review by animesh-g — APPROVED** [review]
